<h2 align='center' style='color:blue'>Dropout Regularization In Deep Neural Network</h2>

This is a dataset that describes sonar chirp returns bouncing off different services. The 60 input variables are the strength of the returns at different angles. It is a **binary classification problem** that requires a model to differentiate rocks from metal cylinders.

Dataset information: https://archive.ics.uci.edu/ml/datasets/Connectionist+Bench+(Sonar,+Mines+vs.+Rocks)
Download it from here: https://archive.ics.uci.edu/ml/machine-learning-databases/undocumented/connectionist-bench/sonar/sonar.all-data

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
import warnings
warnings.filterwarnings('ignore')

In [3]:
df = pd.read_csv("./sonar_dataset.csv", header=None)
df.sample(5)

0       1       2       3       4       5       6       7       8   \
49   0.0119  0.0582  0.0623  0.0600  0.1397  0.1883  0.1422  0.1447  0.0487   
119  0.0261  0.0266  0.0223  0.0749  0.1364  0.1513  0.1316  0.1654  0.1864   
145  0.0721  0.1574  0.1112  0.1085  0.0666  0.1800  0.1108  0.2794  0.1408   
77   0.0336  0.0294  0.0476  0.0539  0.0794  0.0804  0.1136  0.1228  0.1235   
96   0.0181  0.0146  0.0026  0.0141  0.0421  0.0473  0.0361  0.0741  0.1398   

         9   ...      51      52      53      54      55      56      57  \
49   0.0864  ...  0.0025  0.0103  0.0074  0.0123  0.0069  0.0076  0.0073   
119  0.2013  ...  0.0222  0.0175  0.0127  0.0022  0.0124  0.0054  0.0021   
145  0.0795  ...  0.0309  0.0127  0.0082  0.0281  0.0117  0.0092  0.0147   
77   0.0842  ...  0.0150  0.0111  0.0032  0.0035  0.0169  0.0137  0.0015   
96   0.1045  ...  0.0223  0.0255  0.0145  0.0233  0.0041  0.0018  0.0048   

         58      59  60  
49   0.0030  0.0138   R  
119  0.0028  0.0023   M  
145  0.0157  0.0129   M  
77   0.0069  0.0051   R  
96   0.0089  0.0085   R  

[5 rows x 61 columns]

In [4]:
df.shape

(208, 61)

In [5]:
# check for nan values
df.isna()

0      1      2      3      4      5      6      7      8      9   \
0    False  False  False  False  False  False  False  False  False  False   
1    False  False  False  False  False  False  False  False  False  False   
2    False  False  False  False  False  False  False  False  False  False   
3    False  False  False  False  False  False  False  False  False  False   
4    False  False  False  False  False  False  False  False  False  False   
..     ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
203  False  False  False  False  False  False  False  False  False  False   
204  False  False  False  False  False  False  False  False  False  False   
205  False  False  False  False  False  False  False  False  False  False   
206  False  False  False  False  False  False  False  False  False  False   
207  False  False  False  False  False  False  False  False  False  False   

     ...     51     52     53     54     55     56     57     58     59     60  
0    ...  False  False  False  False  False  False  False  False  False  False  
1    ...  False  False  False  False  False  False  False  False  False  False  
2    ...  False  False  False  False  False  False  False  False  False  False  
3    ...  False  False  False  False  False  False  False  False  False  False  
4    ...  False  False  False  False  False  False  False  False  False  False  
..   ...    ...    ...    ...    ...    ...    ...    ...    ...    ...    ...  
203  ...  False  False  False  False  False  False  False  False  False  False  
204  ...  False  False  False  False  False  False  False  False  False  False  
205  ...  False  False  False  False  False  False  False  False  False  False  
206  ...  False  False  False  False  False  False  False  False  False  False  
207  ...  False  False  False  False  False  False  False  False  False  False  

[208 rows x 61 columns]

In [6]:
df.columns

Int64Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
            17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
            34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
            51, 52, 53, 54, 55, 56, 57, 58, 59, 60],
           dtype='int64')

In [7]:
df[60].value_counts() # label is not skewed

M    111
R     97
Name: 60, dtype: int64

In [8]:
X = df.drop(60, axis=1)
y = df[60]
y.head()

0    R
1    R
2    R
3    R
4    R
Name: 60, dtype: object

In [9]:
y = pd.get_dummies(y, drop_first=True)
y.sample(5) # R --> 1 and M --> 0

R
156  0
201  0
57   1
158  0
56   1

In [10]:
y.value_counts()

R
0    111
1     97
dtype: int64

In [11]:
X.head()

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      50      51      52      53      54      55      56  \
0  0.2111  ...  0.0232  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180   
1  0.2872  ...  0.0125  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140   
2  0.6194  ...  0.0033  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316   
3  0.1264  ...  0.0241  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050   
4  0.4459  ...  0.0156  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072   

       57      58      59  
0  0.0084  0.0090  0.0032  
1  0.0049  0.0052  0.0044  
2  0.0164  0.0095  0.0078  
3  0.0044  0.0040  0.0117  
4  0.0048  0.0107  0.0094  

[5 rows x 60 columns]

In [12]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

In [13]:
X_train.head()

0       1       2       3       4       5       6       7       8   \
67   0.0368  0.0403  0.0317  0.0293  0.0820  0.1342  0.1161  0.0663  0.0155   
14   0.0124  0.0433  0.0604  0.0449  0.0597  0.0355  0.0531  0.0343  0.1052   
164  0.0163  0.0198  0.0202  0.0386  0.0752  0.1444  0.1487  0.1484  0.2442   
179  0.0394  0.0420  0.0446  0.0551  0.0597  0.1416  0.0956  0.0802  0.1618   
19   0.0126  0.0149  0.0641  0.1732  0.2565  0.2559  0.2947  0.4110  0.4983   

         9   ...      50      51      52      53      54      55      56  \
67   0.0506  ...  0.0058  0.0091  0.0160  0.0160  0.0081  0.0070  0.0135   
14   0.2120  ...  0.0078  0.0083  0.0057  0.0174  0.0188  0.0054  0.0114   
164  0.2822  ...  0.0027  0.0077  0.0026  0.0031  0.0083  0.0020  0.0084   
179  0.2558  ...  0.0118  0.0146  0.0040  0.0114  0.0032  0.0062  0.0101   
19   0.5920  ...  0.0153  0.0092  0.0035  0.0098  0.0121  0.0006  0.0181   

         57      58      59  
67   0.0067  0.0078  0.0068  
14   0.0196  0.0147  0.0062  
164  0.0108  0.0083  0.0033  
179  0.0068  0.0053  0.0087  
19   0.0094  0.0116  0.0063  

[5 rows x 60 columns]

# Using Deep Learning Model

## Model without Dropout Layer

In [14]:
import tensorflow as tf
from tensorflow import keras

In [33]:
model = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dense(1, activation='sigmoid')
])

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

model.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 0s 761us/step - loss: 0.7361 - accuracy: 0.4625
Epoch 2/100
20/20 [==============================] - 0s 901us/step - loss: 0.6822 - accuracy: 0.5625
Epoch 3/100
20/20 [==============================] - 0s 686us/step - loss: 0.6656 - accuracy: 0.6380
Epoch 4/100
20/20 [==============================] - 0s 852us/step - loss: 0.6555 - accuracy: 0.7117
Epoch 5/100
20/20 [==============================] - 0s 838us/step - loss: 0.6214 - accuracy: 0.7243
Epoch 6/100
20/20 [==============================] - 0s 699us/step - loss: 0.6017 - accuracy: 0.7563
Epoch 7/100
20/20 [==============================] - 0s 709us/step - loss: 0.5866 - accuracy: 0.7077
Epoch 8/100
20/20 [==============================] - 0s 708us/step - loss: 0.5294 - accuracy: 0.8272
Epoch 9/100
20/20 [==============================] - 0s 707us/step - loss: 0.4941 - accuracy: 0.7848
Epoch 10/100
20/20 [==============================] - 0s 698us/step - loss: 0.4455 - accura

20/20 [==============================] - 0s 687us/step - loss: 0.0126 - accuracy: 1.0000
Epoch 82/100
20/20 [==============================] - 0s 766us/step - loss: 0.0147 - accuracy: 1.0000
Epoch 83/100
20/20 [==============================] - 0s 704us/step - loss: 0.0106 - accuracy: 1.0000
Epoch 84/100
20/20 [==============================] - 0s 698us/step - loss: 0.0127 - accuracy: 1.0000
Epoch 85/100
20/20 [==============================] - 0s 708us/step - loss: 0.0074 - accuracy: 1.0000
Epoch 86/100
20/20 [==============================] - 0s 705us/step - loss: 0.0085 - accuracy: 1.0000
Epoch 87/100
20/20 [==============================] - 0s 707us/step - loss: 0.0075 - accuracy: 1.0000
Epoch 88/100
20/20 [==============================] - 0s 705us/step - loss: 0.0072 - accuracy: 1.0000
Epoch 89/100
20/20 [==============================] - 0s 713us/step - loss: 0.0073 - accuracy: 1.0000
Epoch 90/100
20/20 [==============================] - 0s 721us/step - loss: 0.0075 - accuracy: 

In [34]:
model.evaluate(X_test, y_test)

2/2 [==============================] - 0s 2ms/step - loss: 1.0043 - accuracy: 0.7692


[1.0042973756790161, 0.7692307829856873]

Training Accuracy >>> Test Accuracy

In [35]:
y_pred = model.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

[3.2958621e-06 9.9012470e-01 9.9344057e-01 3.4743547e-04 1.0000000e+00
 9.9999923e-01 3.0922061e-01 1.0000000e+00 7.4793672e-05 1.0000000e+00]
[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [36]:
y_test[:10]

R
186  0
155  0
165  0
200  0
58   1
34   1
151  0
18   1
202  0
62   1

In [37]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.74      0.85      0.79        27
           1       0.81      0.68      0.74        25

    accuracy                           0.77        52
   macro avg       0.78      0.77      0.77        52
weighted avg       0.77      0.77      0.77        52



### Model with Dropout Layer

In [38]:
modeld = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(1, activation='sigmoid')
])

modeld.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

modeld.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 0s 889us/step - loss: 0.6958 - accuracy: 0.5264
Epoch 2/100
20/20 [==============================] - 0s 1ms/step - loss: 0.7313 - accuracy: 0.4762
Epoch 3/100
20/20 [==============================] - 0s 850us/step - loss: 0.6909 - accuracy: 0.5443
Epoch 4/100
20/20 [==============================] - 0s 877us/step - loss: 0.6776 - accuracy: 0.6560
Epoch 5/100
20/20 [==============================] - 0s 742us/step - loss: 0.7509 - accuracy: 0.4549
Epoch 6/100
20/20 [==============================] - 0s 728us/step - loss: 0.6832 - accuracy: 0.5273
Epoch 7/100
20/20 [==============================] - 0s 725us/step - loss: 0.7150 - accuracy: 0.5002
Epoch 8/100
20/20 [==============================] - 0s 729us/step - loss: 0.6783 - accuracy: 0.5241
Epoch 9/100
20/20 [==============================] - 0s 730us/step - loss: 0.6410 - accuracy: 0.6233
Epoch 10/100
20/20 [==============================] - 0s 747us/step - loss: 0.6757 - accuracy

20/20 [==============================] - 0s 754us/step - loss: 0.3184 - accuracy: 0.8668
Epoch 82/100
20/20 [==============================] - 0s 767us/step - loss: 0.3670 - accuracy: 0.8579
Epoch 83/100
20/20 [==============================] - 0s 703us/step - loss: 0.2646 - accuracy: 0.8739
Epoch 84/100
20/20 [==============================] - 0s 723us/step - loss: 0.2919 - accuracy: 0.8977
Epoch 85/100
20/20 [==============================] - 0s 697us/step - loss: 0.3269 - accuracy: 0.9037
Epoch 86/100
20/20 [==============================] - 0s 699us/step - loss: 0.3328 - accuracy: 0.9027
Epoch 87/100
20/20 [==============================] - 0s 718us/step - loss: 0.2783 - accuracy: 0.8952
Epoch 88/100
20/20 [==============================] - 0s 755us/step - loss: 0.4277 - accuracy: 0.8299
Epoch 89/100
20/20 [==============================] - 0s 674us/step - loss: 0.3565 - accuracy: 0.8170
Epoch 90/100
20/20 [==============================] - 0s 679us/step - loss: 0.3364 - accuracy: 

In [39]:
modeld.evaluate(X_test, y_test)

2/2 [==============================] - 0s 1ms/step - loss: 0.5002 - accuracy: 0.7500


[0.5002362132072449, 0.75]

Training Accuracy is still good but Test Accuracy Improved

In [40]:
len(y_test)

52

In [41]:
y_pred = modeld.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

[3.1918287e-04 7.7032864e-01 9.2788899e-01 1.1412531e-02 9.9964142e-01
 8.8755202e-01 2.5484258e-01 9.9937868e-01 1.2271494e-02 9.9973977e-01]
[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [42]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.71      0.89      0.79        27
           1       0.83      0.60      0.70        25

    accuracy                           0.75        52
   macro avg       0.77      0.74      0.74        52
weighted avg       0.77      0.75      0.74        52



**You can see that by using dropout layer test accuracy increased from 0.77 to 0.81**